File: exercises/A22/draft/fishSurvival/python/survivalFish.py
def survivalFish(fish_array):
    print(f"fish_array: {fish_array}")
    survivals = [fish_array.pop(0)]

    while len(fish_array) > 0:
        contender = fish_array.pop(0)
        print(f"contender: {contender}")
        while len(survivals) > 0:
            defender = survivals.pop()
            print(f"\tdefender: {defender}")
            if defender > 0 > contender:
                print("\t\tfighting {defender} vs {contender}")
                if defender < -contender:
                    if len(survivals) > 0:
                        continue
                    else:
                        survivals.append(contender)
                        break
                else:
                    survivals.append(defender)
                    break
            else:
                survivals.append(defender)
                survivals.append(contender)
                break
        print(f"\tsurvivals: {survivals}")
        print(f"\tfish_array: {fish_array}")
    return len(survivals)

File: exercises/solutions/A/a22/survivalFish.ts
/**
 * @file survivalFish.test.ts
 */

/**
 * Function to determine the number of surviving fish
 * @param fishArray - Array of integers representing fish
 * @returns Number of surviving fish
 */
export function survivalFish(fishArray: number[]): number {
    //console.log(`fishArray: ${fishArray}`);
    const survivals: number[] = [fishArray.shift() as number];

    while (fishArray.length > 0) {
        const contender = fishArray.shift() as number;
        //console.log(`contender: ${contender}`);

        while (survivals.length > 0) {
            const defender = survivals.pop() as number;
            //console.log(`\tdefender: ${defender}`);

            if (defender > 0 && contender < 0) {
                //console.log(`\t\tfighting ${defender} vs ${contender}`);
                if (defender < -contender) {
                    if (survivals.length > 0) {
                        continue;
                    } else {
                        survivals.push(contender);
                        break;
                    }
                } else {
                    survivals.push(defender);
                    break;
                }
            } else {
                survivals.push(defender);
                survivals.push(contender);
                break;
            }
        }
        //console.log(`\tsurvivals: ${survivals}`);
        //console.log(`\tfishArray: ${fishArray}`);
    }
    return survivals.length;
}
